glib/Marshaller.cs

```diff
@@ -76,33 +76,25 @@ public static string FilenamePtrToStringGFree (IntPtr ptr)
 			return ret;
 		}
 
-#if HAVE_NET_4_6
-		static bool hasFastGetStringOverload = typeof (System.Text.Encoding).GetMethod ("GetString", new [] { typeof (byte*), typeof (int) }) != null;
-		static string Utf8PtrToStringFast (IntPtr ptr, int len)
-		{
-			unsafe
-			{
-				var p = (byte*)ptr;
-				return System.Text.Encoding.UTF8.GetString (p, len);
-			}
-		}
-#endif
+		[DllImport ("libglib-2.0-0.dll", CallingConvention = CallingConvention.Cdecl)]
+		unsafe static extern char* g_utf8_to_utf16 (IntPtr native_str, IntPtr len, IntPtr items_read, IntPtr items_written, IntPtr error);
+
+		[DllImport ("libglib-2.0-0.dll", CallingConvention = CallingConvention.Cdecl)]
+		unsafe static extern IntPtr g_utf16_to_utf8 (char* native_str, IntPtr len, IntPtr items_read, IntPtr items_written, IntPtr error);
 
 		[DllImport("glibsharpglue-2", CallingConvention=CallingConvention.Cdecl)]
 		static extern UIntPtr glibsharp_strlen (IntPtr mem);
 		public static string Utf8PtrToString (IntPtr ptr) 
 		{
 			if (ptr == IntPtr.Zero)
 				return null;
-
-			int len = (int) (uint)glibsharp_strlen (ptr);
-#if HAVE_NET_4_6
-			if (hasFastGetStringOverload)
-				return Utf8PtrToStringFast (ptr, len);
-#endif
-			byte [] bytes = new byte [len];
-			Marshal.Copy (ptr, bytes, 0, len);
-			return System.Text.Encoding.UTF8.GetString (bytes);
+			unsafe
+			{
+				char *utf16 = g_utf8_to_utf16 (ptr, new IntPtr (-1), IntPtr.Zero, IntPtr.Zero, IntPtr.Zero);
```

Comment:
Maybe we don't care as the strings should all be well formatted at this point :P